### PR TITLE
Create HeatMap as suggested in #213

### PR DESCRIPTION
**Pull request by BibMartin**
Using https://github.com/Leaflet/Leaflet.heat

### CODE REVIEW

**Comment by ocefpaf** [comment]
You are on :fire:  this week!

**Comment by BibMartin** [comment]
:musical_note: Hurray ! It's holy-holyday ! :musical_note:

**Comment by andrewgiessel** [comment]
Oh yahhhh! :+1: :tada:

**Comment by ocefpaf** [comment]
heatmap

File: folium/plugins/heat_map.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    def __init__(self, data, name=None,
+                 min_opacity = .5, max_zoom=18, max_val=1.,
+                 radius=25, blur=15, gradient=None,
+                 overlay = True):
+        """Create a Heatmap layer
+
+        Parameters
+        ----------
+        data : list of points of the form [lat, lng, weight]
+            The point you want to plot.
```
Comment (ocefpaf) [inline folium/plugins/heat_map.py]: Let's support 3D numpy as well.  We can listify them internally.

Question:  In your test example you pass a 2D array and things work.  I guess that we should raise an error instead, no?

File: tests/test_plugins.py [hunk 2]
```diff
@@ -134,3 +134,9 @@ def test_timestamped_geo_json(self):
         mape = folium.Map([47, 3], zoom_start=1)
         mape.add_children(plugins.TimestampedGeoJson(data))
         mape._repr_html_()
+
+    def test_heat_map(self):
+        data = (np.random.normal(size=(100,2))*np.array([[1,1]])+np.array([[48,5]])).tolist()
+        mapa = folium.Map([48., 5.], tiles='stamentoner', zoom_start=6)
+        mapa.add_children(plugins.HeatMap(data))
+        mapa._repr_html_()
```
Comment (ocefpaf) [inline tests/test_plugins.py]: This test only if the HTML is created, correct? I guess that all the plugins are tested like that...  Not sure if we can do better.

File: folium/plugins/heat_map.py (hunk 1 above)
Comment (BibMartin) [inline folium/plugins/heat_map.py]: the js library accepts arrays of shape (n,2) and (n,3) ; so we shall do the same thing. A new commit just do that.

File: tests/test_plugins.py (hunk 2 above)
Comment (BibMartin) [inline tests/test_plugins.py]: Correct.

We can do a bit better in doing the same thing as `test_folium.py` : compare the output of `.render()` with code generated from a validation template.
But this is a heavy procedure and I'm not sure this is needed for all plugins (am I lazy ?).

File: folium/plugins/heat_map.py (hunk 1 above)
Comment (ocefpaf) [inline folium/plugins/heat_map.py]: > the js library accepts arrays of shape (n,2) and (n,3) ; so we shall do the same thing.

OK.  If the JS support let's do that then... But what is the reason?  I confess I just glimpsed over the docs, but I could not find a rational for that.

File: tests/test_plugins.py (hunk 2 above)
Comment (ocefpaf) [inline tests/test_plugins.py]: > But this is a heavy procedure and I'm not sure this is needed for all plugins (am I lazy ?).

Heavy?  Yes.
Are you lazy?  No!
Should someone else do that? (Me :unamused:)

**Comment by ocefpaf** [comment]
It is all good to me.  Just a minor typo and then good to go!

Thanks for the great work @BibMartin!

**Comment by BibMartin** [comment]
typo fixed ; commits squashed ...

**Comment by ocefpaf** [comment]
And merged!

**Comment by jackboot7** [comment]
Amazing work @BibMartin, just used your HeatMap plugin to generate a pretty map of geolocalized data.

Thank you!

**Comment by ocefpaf** [comment]
Hi @jackboot7 consider sending your notebook to our ["gallery."](https://github.com/python-visualization/folium/tree/master/examples)

**Comment by jackboot7** [comment]
@ocefpaf: Sure, I'll do it as soon as I'm done with my program.

**Comment by ocefpaf** [comment]
Thanks :+1:

**Comment by jackboot7** [comment]
@ocefpaf How would you prefer that I share my usage of Folium with you, guys? I made a blog post about the my use case (http://www.jackboot7.com/visualizing-tweets.html )

**Comment by ocefpaf** [comment]
The blog reference is fine.  We don't have a list of use cases right now but I want to start one.

However, if you find the time it would be nice to get a PR with that notebook in our examples folder:

https://github.com/python-visualization/folium/tree/master/examples

**Pull request merged**